flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/BatchTableDescriptor.scala

```diff
@@ -0,0 +1,126 @@
+import org.apache.flink.table.factories.{BatchTableSinkFactory, BatchTableSourceFactory, TableFactoryService}
+
+/**
+  * Descriptor for specifying a table source and/or sink in a batch environment.
+  */
+class BatchTableDescriptor(
+    private val tableEnv: BatchTableEnvironment,
+    private val connectorDescriptor: ConnectorDescriptor)
+  extends TableDescriptor
+  with SchematicDescriptor
+  with RegistrableDescriptor {
+
+  private var formatDescriptor: Option[FormatDescriptor] = None
+  private var schemaDescriptor: Option[Schema] = None
+
+  /**
+    * Internal method for properties conversion.
+    */
+  override private[flink] def addProperties(properties: DescriptorProperties): Unit = {
+    connectorDescriptor.addProperties(properties)
+    formatDescriptor.foreach(_.addProperties(properties))
+    schemaDescriptor.foreach(_.addProperties(properties))
+  }
+
+  /**
+    * Searches for the specified table source, configures it accordingly, and registers it as
+    * a table under the given name.
+    *
+    * @param name table name to be registered in the table environment
+    */
+  override def registerTableSource(name: String): Unit = {
+    val javaMap = getValidProperties.asMap
+    val tableSource = TableFactoryService
+      .find(classOf[BatchTableSourceFactory[_]], javaMap)
+      .createBatchTableSource(javaMap)
+    tableEnv.registerTableSource(name, tableSource)
+  }
+
+  /**
+    * Searches for the specified table sink, configures it accordingly, and registers it as
+    * a table under the given name.
+    *
+    * @param name table name to be registered in the table environment
+    */
+  override def registerTableSink(name: String): Unit = {
+    val javaMap = getValidProperties.asMap
+    val tableSink = TableFactoryService
+      .find(classOf[BatchTableSinkFactory[_]], javaMap)
+      .createBatchTableSink(javaMap)
+    tableEnv.registerTableSink(name, tableSink)
+  }
+
+  /**
+    * Searches for the specified table source and sink, configures them accordingly, and registers
+    * them as a table under the given name.
+    *
+    * @param name table name to be registered in the table environment
+    */
+  override def registerTableSourceAndSink(name: String): Unit = {
+    registerTableSource(name)
+    registerTableSink(name)
+  }
+
+  /**
+    * Specifies the format that defines how to read data from a connector.
+    */
+  override def withFormat(format: FormatDescriptor): BatchTableDescriptor = {
+    formatDescriptor = Some(format)
+    this
+  }
+
+  /**
+    * Specifies the resulting table schema.
+    */
+  override def withSchema(schema: Schema): BatchTableDescriptor = {
+    schemaDescriptor = Some(schema)
+    this
+  }
+
+  override def toString: String = {
+    getValidProperties.toString
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  private def getValidProperties: DescriptorProperties = {
```

Comment:
duplicated code with `StreamTableDescriptor. getValidProperties`

Reply:
I agree but I cannot move it up the class hierarchy because it would be public in Java. I will create a util class.